File: frontend/src/lib/types/schedule.ts
import type { User } from './user';

export type RotationType = 'daily' | 'weekly' | 'custom';

export interface Schedule {
	id: string;
	organization_id: string;
	team_id?: string;
	name: string;
	description?: string;
	timezone: string;
	created_at: string;
	updated_at: string;
}

export interface ScheduleRotation {
	id: string;
	schedule_id: string;
	name: string;
	rotation_type: RotationType;
	rotation_length: number;
	start_date: string;
	start_time: string;
	end_time?: string;
	handoff_day?: number;
	handoff_time: string;
	created_at: string;
	updated_at: string;
}

export interface ScheduleRotationParticipant {
	id: string;
	rotation_id: string;
	user_id: string;
	position: number;
	created_at: string;
}

export interface ParticipantWithUser extends ScheduleRotationParticipant {
	user: User;
}

export interface ScheduleOverride {
	id: string;
	schedule_id: string;
	user_id: string;
	start_time: string;
	end_time: string;
	note?: string;
	created_at: string;
	updated_at: string;
}

export interface OnCallUser {
	user_id: string;
	user?: User;
	schedule_id: string;
	start_time: string;
	end_time: string;
	is_override: boolean;
}

export interface ScheduleWithRotations extends Schedule {
	rotations: ScheduleRotation[];
}

export interface RotationWithParticipants extends ScheduleRotation {
	participants: ParticipantWithUser[];
}

// Request types

export interface CreateScheduleRequest {
	team_id?: string;
	name: string;
	description?: string;
	timezone?: string;
}

export interface UpdateScheduleRequest {
	team_id?: string;
	name?: string;
	description?: string;
	timezone?: string;
}

export interface CreateRotationRequest {
	name: string;
	rotation_type: RotationType;
	rotation_length: number;
	start_date: string;
	start_time?: string;
	end_time?: string;
	handoff_day?: number;
	handoff_time?: string;
}

export interface UpdateRotationRequest {
	name?: string;
	rotation_type?: RotationType;
	rotation_length?: number;
	start_date?: string;
	start_time?: string;
	end_time?: string;
	handoff_day?: number;
	handoff_time?: string;
}

export interface AddParticipantRequest {
	user_id: string;
	position: number;
}

export interface ReorderParticipantsRequest {
	user_ids: string[];
}

export interface CreateOverrideRequest {
	user_id: string;
	start_time: string;
	end_time: string;
	note?: string;
}

export interface UpdateOverrideRequest {
	user_id?: string;
	start_time?: string;
	end_time?: string;
	note?: string;
}

// Response types

export interface ListSchedulesResponse {
	schedules: Schedule[];
}

export interface ListRotationsResponse {
	rotations: ScheduleRotation[];
}

export interface ListParticipantsResponse {
	participants: ParticipantWithUser[];
}

export interface ListOverridesResponse {
	overrides: ScheduleOverride[];
}

File: backend/internal/core/dto/schedule.go
package dto

import (
	"github.com/google/uuid"
)

type CreateScheduleRequest struct {
	TeamID      *uuid.UUID `json:"team_id"`
	Name        string     `json:"name" binding:"required"`
	Description *string    `json:"description"`
	Timezone    string     `json:"timezone"`
}

type UpdateScheduleRequest struct {
	TeamID      *uuid.UUID `json:"team_id"`
	Name        *string    `json:"name"`
	Description *string    `json:"description"`
	Timezone    *string    `json:"timezone"`
}

type CreateRotationRequest struct {
	Name           string  `json:"name" binding:"required"`
	RotationType   string  `json:"rotation_type" binding:"required"`
	RotationLength int     `json:"rotation_length" binding:"required"`
	StartDate      string  `json:"start_date" binding:"required"`
	StartTime      string  `json:"start_time"`
	EndTime        *string `json:"end_time"`
	HandoffDay     *int    `json:"handoff_day"`
	HandoffTime    string  `json:"handoff_time"`
}

type UpdateRotationRequest struct {
	Name           *string `json:"name"`
	RotationType   *string `json:"rotation_type"`
	RotationLength *int    `json:"rotation_length"`
	StartDate      *string `json:"start_date"`
	StartTime      *string `json:"start_time"`
	EndTime        *string `json:"end_time"`
	HandoffDay     *int    `json:"handoff_day"`
	HandoffTime    *string `json:"handoff_time"`
}

type AddParticipantRequest struct {
	UserID   uuid.UUID `json:"user_id" binding:"required"`
	Position int       `json:"position" binding:"min=0"`
}

type ReorderParticipantsRequest struct {
	UserIDs []uuid.UUID `json:"user_ids" binding:"required"`
}

type CreateOverrideRequest struct {
	UserID    uuid.UUID `json:"user_id" binding:"required"`
	StartTime string    `json:"start_time" binding:"required"`
	EndTime   string    `json:"end_time" binding:"required"`
	Note      *string   `json:"note"`
}

type UpdateOverrideRequest struct {
	UserID    *uuid.UUID `json:"user_id"`
	StartTime *string    `json:"start_time"`
	EndTime   *string    `json:"end_time"`
	Note      *string    `json:"note"`
}
